File: benchmarks/fixtures/webapp_py/models/session.py
"""Session model."""

from typing import Optional
from datetime import datetime, timedelta

from models.user import User


class Session:
    """Represents an authentication session."""

    _store = {}

    def __init__(self, token: str, user: User, expires_at: datetime):
        self.token = token
        self.user = user
        self.expires_at = expires_at

    @classmethod
    def create(cls, user: User, token: str, expires_in: int = 3600) -> "Session":
        """Create a new session."""
        expires_at = datetime.utcnow() + timedelta(seconds=expires_in)
        session = cls(token=token, user=user, expires_at=expires_at)
        cls._store[token] = session
        return session

    @classmethod
    def find_by_token(cls, token: str) -> Optional["Session"]:
        """Look up a session by token."""
        return cls._store.get(token)

    @classmethod
    def find_all_by_user(cls, user: User) -> list:
        """Find all sessions for a user."""
        return [s for s in cls._store.values() if s.user.id == user.id]

    def delete(self):
        """Delete this session."""
        self._store.pop(self.token, None)

    def is_expired(self) -> bool:
        """Check if this session has expired."""
        return self.expires_at < datetime.utcnow()

File: benchmarks/fixtures/webapp_rs/models/session.rs
use crate::models::user::User;

pub struct Session {
    pub token: String,
    pub user_id: u64,
    pub expires_at: u64,
}

impl Session {
    pub fn create(user: &User, token: String, expires_in: u64) -> Self {
        let expires_at = current_timestamp() + expires_in;
        Self {
            token,
            user_id: user.id,
            expires_at,
        }
    }

    pub fn find_by_token(token: &str) -> Option<Session> {
        println!("Looking up session: {token}");
        None
    }

    pub fn find_all_by_user(user: &User) -> Vec<Session> {
        println!("Looking up sessions for user: {}", user.id);
        vec![]
    }

    pub fn delete(&self) {
        println!("Deleting session: {}", self.token);
    }

    pub fn is_expired(&self) -> bool {
        self.expires_at < current_timestamp()
    }
}

fn current_timestamp() -> u64 {
    0
}
